### PR TITLE
Add access mechanism for state vectors in ODE solvers [ode-state-io-dev]

### PR DESCRIPTION
**Pull request by IdoAkkerman**
This PR adds some simple access functions for state/history vectors in the ODE solvers.
For generalized alpha this is the rate, while for the AB and AM it are the rates of several previous timesteps.

The mechanisms are also added to the ode unit tests.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1430](https://github.com/mfem/mfem/pull/1430) | @IdoAkkerman | @v-dobrev | @cjvogl + @stefanozampini | 04/29/20 | 07/07/20 | 07/15/20 | |

### CODE REVIEW

**Comment by IdoAkkerman** [comment]
1. I added the suggested interface and included it in the unit test.

2. I changed some virtual statements in ode to override.

**Comment by stefanozampini** [comment]
Should the interface for sundials and PETSc change using override for Init and Step too?

File: linalg/ode.hpp [hunk 1]
```diff
@@ -114,9 +118,9 @@ class ForwardEulerSolver : public ODESolver
    Vector dxdt;
 
 public:
-   virtual void Init(TimeDependentOperator &_f);
+   void Init(TimeDependentOperator &_f) override;
```
Comment (cjvogl) [inline linalg/ode.hpp:121]: Did you mean to remove the `virtual` keyword here?  In other words, is it the case that a class that derives the `ForwardEulerSolver` should no longer be able to redefine the `Init` function?

Comment (cjvogl) [reply linalg/ode.hpp:121]: Oops, I think I goofed here.  I believe that `Init` is still considered `virtual` even without the `virtual` keyword... I'm checking on that now.

Comment (cjvogl) [reply linalg/ode.hpp:121]: Yup, my mistake: `Init` is still considered `virtual`.  Please disregard my comment.

File: linalg/ode.hpp [hunk 2]
```diff
@@ -91,6 +91,22 @@ class ODESolver
...
    }
 
+   /// Function for getting and setting the state vectors
+   virtual int   GetMaxStateSize() { return 0; }
+   virtual int   GetStateSize() { return 0; }
+   virtual void  GetStateVector(int i, Vector &state)
+   {
+      mfem_error("ODESolver has no state vectors");
+   }
+   virtual const Vector &GetStateVector(int i)
```
Comment (cjvogl) [inline linalg/ode.hpp:101]: Appveyor is upset that this function doesn't return anything.

**Comment by cjvogl** [comment]
> 
> 
> Should the interface for sundials and PETSc change using override for Init and Step too?

Personally, I think that it is good practice to use the `override` keyword for any virtual function definition that overrides a base class definition.  That said, the keyword is not used in much (any?) of the MFEM code, so I wouldn't want to ask @IdoAkkerman to address that.

**Comment by v-dobrev** [comment]
I agree that using `override` is good for clarity.

It is not used in many places mainly because it is c++11 feature and we have not made the effort to add it after the transition to c++11.

File: linalg/ode.hpp (hunk 2 above)
Comment (v-dobrev) [reply linalg/ode.hpp:101]: We probably can just return an empty temporary here: something like `return Vector();` or `Vector s; return s;`

Comment (IdoAkkerman) [reply linalg/ode.hpp:101]: In both cases you get a warning  

> -Wreturn-local-addr

I went with `Vector *s = NULL; return *s;`

**Comment by IdoAkkerman** [comment]
@cjvogl I was unaware of the override functionality. Thank you for pointing me to it. To answer your question. I did purposly change more virtuals in the hpp file. Not all as I am a bit uncertain regarding the destructors.

I had to check myself, but I also came to the conclusion that an overriden function van be overriden later on. So using the override statement does not alter any functionallity, but adds a layer of protection/checking.

File: tests/unit/linalg/test_ode2.cpp [hunk 3]
```diff
@@ -112,18 +125,49 @@ TEST_CASE("Second order ODE methods",
          std::cout<<std::setw(12)<<err_u[0]
                   <<std::setw(12)<<err_du[0]<<std::endl;
 
+         Array<Vector> uh(ode_solver->GetMaxStateSize());
```
Comment (v-dobrev) [inline tests/unit/linalg/test_ode2.cpp:128]: The `mfem::Array` class only supports POD types, so this line should use something like `std::vector<Vector>` instead.

File: tests/unit/linalg/test_ode.cpp [hunk 4]
```diff
@@ -113,18 +125,48 @@ TEST_CASE("First order ODE methods",
                   <<std::setw(12)<<"Order"<<std::endl;
          std::cout<<std::setw(12)<<err[0]<<std::endl;
 
-         for (int l = 1; l< levels; l++)
+         Array<Vector> uh(ode_solver->GetMaxStateSize());
```
Comment (v-dobrev) [inline tests/unit/linalg/test_ode.cpp:128]: `Array<Vector>` -> `std::vector<Vector>`

**Comment by v-dobrev** [comment]
> Should the interface for sundials and PETSc change using override for Init and Step too?

@stefanozampini, we can do that in another future PR.

Other than that do you have any feedback on the PR?

File: tests/unit/linalg/test_ode2.cpp (hunk 3 above)
Comment (IdoAkkerman) [reply tests/unit/linalg/test_ode2.cpp:128]: I had to google POD. But I guess it stands for plain-old-data, and assumes all the data is contigous in memory, or something like that, such that memcpy (etc) works... Just to make sure I understand and in an attempt to not make the mistake in the future.

File: tests/unit/linalg/test_ode.cpp (hunk 4 above)
Comment (IdoAkkerman) [reply tests/unit/linalg/test_ode.cpp]: Made the change.

File: tests/unit/linalg/test_ode2.cpp (hunk 3 above)
Comment (IdoAkkerman) [reply tests/unit/linalg/test_ode2.cpp]: Made the change.

Comment (v-dobrev) [reply tests/unit/linalg/test_ode2.cpp]: Here is a precise definition: https://en.cppreference.com/w/cpp/named_req/PODType.

The main reason for the POD requirement in `Array` is that we do not call contructors or destructors for the array entries.

**Comment by v-dobrev** [comment]
ping: @stefanozampini

**Comment by stefanozampini** [comment]
I thought I already approved.  How can I approve? I dont see a button to do that

**Comment by v-dobrev** [comment]
LGTM

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Pull request merged**